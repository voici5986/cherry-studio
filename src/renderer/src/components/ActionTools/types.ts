/**
 * 动作工具基本信息
 */
export interface ActionToolSpec {
  id: string
  type: 'core' | 'quick'
  order: number
}

/**
 * 动作工具定义接口
 * @param id 唯一标识符
 * @param type 工具类型
 * @param order 显示顺序，越小越靠右
 * @param icon 按钮图标
 * @param tooltip 提示文本
 * @param visible 显示条件
 * @param onClick 点击动作
 * @param children 子工具（例如 more 下拉菜单）
 */
export interface ActionTool extends ActionToolSpec {
  icon: React.ReactNode
  tooltip?: string
  visible?: () => boolean
  onClick?: () => void
  children?: Omit<ActionTool, 'children'>[]
}

/**
 * 子组件向父组件注册工具所需的 props
 */
export interface ToolRegisterProps {
  setTools?: (value: React.SetStateAction<ActionTool[]>) => void
}
