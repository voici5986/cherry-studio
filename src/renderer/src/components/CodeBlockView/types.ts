/**
 * 视图模式
 */
export type ViewMode = 'source' | 'special' | 'split'
