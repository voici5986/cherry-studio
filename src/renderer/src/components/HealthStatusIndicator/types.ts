import { HealthStatus } from '@renderer/types/healthCheck'

/**
 * 用于展示单个健康检查结果的必要数据
 */
export interface HealthResult {
  status: HealthStatus
  latency?: number
  error?: string
  // 用于在 Tooltip 中显示额外上下文信息，例如 API Key 或模型名称
  label?: string
}
