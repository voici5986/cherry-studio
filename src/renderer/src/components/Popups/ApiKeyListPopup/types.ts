import { PreprocessProvider, Provider, WebSearchProvider } from '@renderer/types'

/**
 * API key 格式有效性
 */
export type ApiKeyValidity = {
  isValid: boolean
  error?: string
}

export type ApiProvider = Provider | WebSearchProvider | PreprocessProvider

export type UpdateProviderFunc = (p: Partial<Provider>) => void

export type UpdateWebSearchProviderFunc = (p: Partial<WebSearchProvider>) => void

export type UpdatePreprocessProviderFunc = (p: Partial<PreprocessProvider>) => void

export type UpdateApiProviderFunc = UpdateProviderFunc | UpdateWebSearchProviderFunc | UpdatePreprocessProviderFunc
