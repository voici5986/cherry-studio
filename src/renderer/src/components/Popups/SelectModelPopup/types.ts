import { Model } from '@renderer/types'
import { ReactNode } from 'react'

/**
 * 滚动触发来源类型
 */
export type ScrollTrigger = 'initial' | 'search' | 'keyboard' | 'none'

/**
 * 列表项分类，组名也作为列表项
 */
export type ListItemType = 'group' | 'model'

/**
 * 扁平化列表项基础类型
 */
export type FlatListBaseItem = {
  key: string
  type: ListItemType
  name: ReactNode
  icon?: ReactNode
  isSelected?: boolean
}

/**
 * 模型分组列表项
 */
export type FlatListGroup = FlatListBaseItem & {
  type: 'group'
  actions?: ReactNode
}

/**
 * 模型列表项
 */
export type FlatListModel = FlatListBaseItem & {
  type: 'model'
  model: Model
  tags?: ReactNode
  isPinned?: boolean
}

/**
 * 扁平化列表项
 */
export type FlatListItem = FlatListGroup | FlatListModel
