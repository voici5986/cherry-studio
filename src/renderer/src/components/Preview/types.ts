/**
 * 预览组件的基本 props
 */
export interface BasicPreviewProps {
  children: string
  enableToolbar?: boolean
}

/**
 * 通过 useImperativeHandle 暴露的方法类型
 */
export interface BasicPreviewHandles {
  pan: (dx: number, dy: number, absolute?: boolean) => void
  zoom: (delta: number, absolute?: boolean) => void
  copy: () => Promise<void>
  download: (format: 'svg' | 'png') => Promise<void>
}
