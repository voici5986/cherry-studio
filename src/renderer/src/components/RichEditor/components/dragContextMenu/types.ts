import type { Editor } from '@tiptap/core'
import type { Node } from '@tiptap/pm/model'
import type { ReactNode } from 'react'

/**
 * 操作组类型 - 用于菜单项分组
 */
export enum ActionGroup {
  TRANSFORM = 'transform', // 节点转换操作
  FORMAT = 'format', // 格式化操作
  BLOCK = 'block', // 块级操作
  INSERT = 'insert', // 插入操作
  AI = 'ai' // AI 相关操作 (预留)
}

/**
 * 菜单操作项接口
 */
export interface MenuAction {
  /** 操作唯一标识 */
  id: string
  /** 显示标签 */
  label: string
  /** 图标 */
  icon?: ReactNode
  /** 操作组 */
  group: ActionGroup
  /** 快捷键描述 */
  shortcut?: string
  /** 是否为危险操作 */
  danger?: boolean
  /** 是否可用 */
  isEnabled: (editor: Editor, node: Node, pos: number) => boolean
  /** 执行操作 */
  execute: (editor: Editor, node: Node, pos: number) => void
  /** 自定义类名 */
  className?: string
}

/**
 * 节点转换选项
 */
export interface NodeTransformOptions {
  /** 目标节点类型 */
  nodeType: string
  /** 节点属性 */
  attrs?: Record<string, any>
  /** 是否保留内容 */
  preserveContent?: boolean
}

/**
 * 拖拽上下文菜单属性
 */
export interface DragContextMenuProps {
  /** 编辑器实例 */
  editor: Editor
  /** 当前节点 */
  node: Node
  /** 节点在文档中的位置 */
  position: number
  /** 菜单显示状态 */
  visible: boolean
  /** 菜单位置 */
  menuPosition: { x: number; y: number }
  /** 关闭回调 */
  onClose: () => void
  /** 自定义操作 */
  customActions?: MenuAction[]
  /** 禁用的操作 ID 列表 */
  disabledActions?: string[]
}

/**
 * 拖拽手柄属性
 */
export interface DragHandleProps {
  /** 编辑器实例 */
  editor: Editor
  /** 当前节点 */
  node: Node
  /** 节点位置 */
  position: number
  /** 是否显示 */
  visible: boolean
  /** 点击回调 */
  onClick: () => void
  /** 拖拽开始回调 */
  onDragStart?: (e: DragEvent) => void
  /** 自定义类名 */
  className?: string
}

/**
 * 菜单可见性配置
 */
export interface MenuVisibilityOptions {
  /** 当前编辑器实例 */
  editor: Editor
  /** 当前节点 */
  node: Node
  /** 节点位置 */
  position: number
  /** 自定义可见性规则 */
  customRules?: Array<(editor: Editor, node: Node, pos: number) => boolean>
}

/**
 * 菜单操作结果
 */
export interface MenuActionResult {
  /** 是否成功执行 */
  success: boolean
  /** 错误信息 */
  error?: string
  /** 是否需要关闭菜单 */
  shouldCloseMenu?: boolean
}

/**
 * 颜色选择器选项
 */
export interface ColorOption {
  /** 颜色值 */
  color: string
  /** 显示名称 */
  name: string
  /** 是否为默认颜色 */
  isDefault?: boolean
}

/**
 * 节点转换映射
 */
export interface NodeTransformMap {
  [key: string]: {
    /** 显示名称 */
    label: string
    /** 图标 */
    icon: ReactNode
    /** 转换配置 */
    transform: NodeTransformOptions
    /** 是否可用的检查函数 */
    isAvailable?: (editor: Editor, currentNode: Node) => boolean
  }
}

/**
 * 拖拽上下文菜单配置
 */
export interface DragContextMenuConfig {
  /** 是否启用 */
  enabled: boolean
  /** 默认操作列表 */
  defaultActions: string[]
  /** 自定义操作 */
  customActions?: MenuAction[]
  /** 操作组排序 */
  groupOrder?: ActionGroup[]
  /** 颜色选择器配置 */
  colorOptions?: ColorOption[]
  /** 节点转换映射 */
  transformMap?: NodeTransformMap
  /** 菜单样式配置 */
  menuStyles?: {
    maxWidth?: number
    maxHeight?: number
    showShortcuts?: boolean
  }
}

/**
 * 钩子返回值 - useDragContextMenu
 */
export interface UseDragContextMenuReturn {
  /** 菜单是否可见 */
  isMenuVisible: boolean
  /** 菜单位置 */
  menuPosition: { x: number; y: number }
  /** 当前节点信息 */
  currentNode: { node: Node; position: number } | null
  /** 显示菜单 */
  showMenu: (node: Node, position: number, clientPos: { x: number; y: number }) => void
  /** 隐藏菜单 */
  hideMenu: () => void
  /** 执行操作 */
  executeAction: (action: MenuAction) => Promise<MenuActionResult>
}

/**
 * 钩子返回值 - useMenuActionVisibility
 */
export interface UseMenuActionVisibilityReturn {
  /** 可见的操作列表 */
  visibleActions: MenuAction[]
  /** 按组分类的操作 */
  actionsByGroup: Record<ActionGroup, MenuAction[]>
  /** 刷新可见性 */
  refreshVisibility: () => void
}

/**
 * 事件处理器类型
 */
export interface EventHandlers {
  onMenuShow?: (node: Node, position: number) => void
  onMenuHide?: () => void
  onActionExecute?: (action: MenuAction, result: MenuActionResult) => void
  onError?: (error: Error, context: string) => void
}

/**
 * 位置计算选项
 */
export interface PositionOptions {
  /** 偏移量 */
  offset?: { x: number; y: number }
  /** 边界约束 */
  boundary?: HTMLElement
  /** 对齐方式 */
  align?: 'start' | 'center' | 'end'
  /** 自动调整位置 */
  autoAdjust?: boolean
}
