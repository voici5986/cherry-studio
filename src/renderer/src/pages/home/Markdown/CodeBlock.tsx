import { CodeBlockView, HtmlArtifactsCard } from '@renderer/components/CodeBlockView'
import { EVENT_NAMES, EventEmitter } from '@renderer/services/EventService'
import store from '@renderer/store'
import { messageBlocksSelectors } from '@renderer/store/messageBlock'
import { MessageBlockStatus } from '@renderer/types/newMessage'
import { getCodeBlockId, isOpenFenceBlock } from '@renderer/utils/markdown'
import type { Node } from 'mdast'
import React, { memo, useCallback, useMemo } from 'react'

interface Props {
  children: string
  className?: string
  node?: Omit<Node, 'type'>
  blockId: string // Message block id
  [key: string]: any
}

const CodeBlock: React.FC<Props> = ({ children, className, node, blockId }) => {
  const languageMatch = /language-([\w-+]+)/.exec(className || '')
  const isMultiline = children?.includes('\n')
  const language = languageMatch?.[1] ?? (isMultiline ? 'text' : null)

  // 代码块 id
  const id = useMemo(() => getCodeBlockId(node?.position?.start), [node?.position?.start])

  // 消息块
  const msgBlock = messageBlocksSelectors.selectById(store.getState(), blockId)
  const isStreaming = useMemo(() => msgBlock?.status === MessageBlockStatus.STREAMING, [msgBlock?.status])

  const handleSave = useCallback(
    (newContent: string) => {
      if (id !== undefined) {
        EventEmitter.emit(EVENT_NAMES.EDIT_CODE_BLOCK, {
          msgBlockId: blockId,
          codeBlockId: id,
          newContent
        })
      }
    },
    [blockId, id]
  )

  if (language !== null) {
    // HTML 代码块特殊处理
    if (language === 'html') {
      const isOpenFence = isOpenFenceBlock(children?.length, languageMatch?.[1]?.length, node?.position)
      return <HtmlArtifactsCard html={children} onSave={handleSave} isStreaming={isStreaming && isOpenFence} />
    }

    return (
      <CodeBlockView language={language} onSave={handleSave}>
        {children}
      </CodeBlockView>
    )
  }

  return (
    <code className={className} style={{ textWrap: 'wrap', fontSize: '95%', padding: '2px 4px' }}>
      {children}
    </code>
  )
}

export default memo(CodeBlock)
