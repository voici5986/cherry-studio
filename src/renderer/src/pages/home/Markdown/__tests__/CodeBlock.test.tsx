import { MessageBlockStatus } from '@renderer/types/newMessage'
import { fireEvent, render, screen } from '@testing-library/react'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import CodeBlock from '../CodeBlock'

// Hoisted mocks
const mocks = vi.hoisted(() => ({
  EventEmitter: {
    emit: vi.fn()
  },
  getCodeBlockId: vi.fn(),
  isOpenFenceBlock: vi.fn(),
  selectById: vi.fn(),
  CodeBlockView: vi.fn(({ onSave, children }) => (
    <div>
      <code>{children}</code>
      <button type="button" onClick={() => onSave('new code content')}>
        Save
      </button>
    </div>
  )),
  HtmlArtifactsCard: vi.fn(({ onSave, html }) => (
    <div>
      <div>{html}</div>
      <button type="button" onClick={() => onSave('new html content')}>
        Save HTML
      </button>
    </div>
  ))
}))

// Mock modules
vi.mock('@renderer/services/EventService', () => ({
  EVENT_NAMES: { EDIT_CODE_BLOCK: 'EDIT_CODE_BLOCK' },
  EventEmitter: mocks.EventEmitter
}))

vi.mock('@renderer/utils/markdown', () => ({
  getCodeBlockId: mocks.getCodeBlockId,
  isOpenFenceBlock: mocks.isOpenFenceBlock
}))

vi.mock('@renderer/store', () => ({
  default: {
    getState: vi.fn(() => ({})) // Mock store, state doesn't matter here
  }
}))

vi.mock('@renderer/store/messageBlock', () => ({
  messageBlocksSelectors: {
    selectById: mocks.selectById
  }
}))

vi.mock('@renderer/components/CodeBlockView', () => ({
  CodeBlockView: mocks.CodeBlockView,
  HtmlArtifactsCard: mocks.HtmlArtifactsCard
}))

describe('CodeBlock', () => {
  const defaultProps = {
    blockId: 'test-msg-block-id',
    node: {
      position: {
        start: { line: 1, column: 1, offset: 0 },
        end: { line: 2, column: 1, offset: 2 },
        value: 'console.log("hello world")'
      }
    },
    children: 'console.log("hello world")',
    className: 'language-javascript'
  }

  beforeEach(() => {
    vi.clearAllMocks()
    // Default mock return values
    mocks.getCodeBlockId.mockReturnValue('test-code-block-id')
    mocks.isOpenFenceBlock.mockReturnValue(false)
    mocks.selectById.mockReturnValue({
      id: 'test-msg-block-id',
      status: MessageBlockStatus.SUCCESS
    })
  })

  describe('rendering', () => {
    it('should render a snapshot', () => {
      const { container } = render(<CodeBlock {...defaultProps} />)
      expect(container).toMatchSnapshot()
    })

    it('should render inline code when no language match is found', () => {
      const inlineProps = {
        ...defaultProps,
        className: undefined,
        children: 'inline code'
      }
      render(<CodeBlock {...inlineProps} />)

      const codeElement = screen.getByText('inline code')
      expect(codeElement.tagName).toBe('CODE')
      expect(mocks.CodeBlockView).not.toHaveBeenCalled()
    })
  })

  describe('save', () => {
    it('should call EventEmitter with correct payload when saving a standard code block', () => {
      render(<CodeBlock {...defaultProps} />)

      // Simulate clicking the save button inside the mocked CodeBlockView
      const saveButton = screen.getByText('Save')
      fireEvent.click(saveButton)

      // Verify getCodeBlockId was called
      expect(mocks.getCodeBlockId).toHaveBeenCalledWith(defaultProps.node.position.start)

      // Verify EventEmitter.emit was called
      expect(mocks.EventEmitter.emit).toHaveBeenCalledOnce()
      expect(mocks.EventEmitter.emit).toHaveBeenCalledWith('EDIT_CODE_BLOCK', {
        msgBlockId: 'test-msg-block-id',
        codeBlockId: 'test-code-block-id',
        newContent: 'new code content'
      })
    })

    it('should call EventEmitter with correct payload when saving an HTML block', () => {
      const htmlProps = {
        ...defaultProps,
        className: 'language-html',
        children: '<h1>Hello</h1>'
      }
      render(<CodeBlock {...htmlProps} />)

      // Simulate clicking the save button inside the mocked HtmlArtifactsCard
      const saveButton = screen.getByText('Save HTML')
      fireEvent.click(saveButton)

      // Verify getCodeBlockId was called
      expect(mocks.getCodeBlockId).toHaveBeenCalledWith(htmlProps.node.position.start)

      // Verify EventEmitter.emit was called
      expect(mocks.EventEmitter.emit).toHaveBeenCalledOnce()
      expect(mocks.EventEmitter.emit).toHaveBeenCalledWith('EDIT_CODE_BLOCK', {
        msgBlockId: 'test-msg-block-id',
        codeBlockId: 'test-code-block-id',
        newContent: 'new html content'
      })
    })
  })
})
