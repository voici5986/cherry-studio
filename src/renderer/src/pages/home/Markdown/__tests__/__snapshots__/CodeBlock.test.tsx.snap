// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CodeBlock > rendering > should render a snapshot 1`] = `
<div>
  <div>
    <code>
      console.log("hello world")
    </code>
    <button
      type="button"
    >
      Save
    </button>
  </div>
</div>
`;
